statistics/table.go

```diff
@@ -596,6 +596,144 @@ func CETraceRange(sctx sessionctx.Context, tableID int64, colNames []string, ran
+	// try to find available stats in column stats
+	if colStats, ok := coll.Columns[uniqueID]; ok && colStats != nil && !colStats.IsInvalid(sctx, coll.Pseudo) {
+		return false, uniqueID
+	}
+	// try to find available stats in single column index stats (except for prefix index)
+	for idxStatsIdx, cols := range coll.Idx2ColumnIDs {
+		if len(cols) == 1 && cols[0] == uniqueID {
+			idxStats, ok := coll.Indices[idxStatsIdx]
+			if ok &&
+				idxStats.Info.Columns[0].Length == types.UnspecifiedLength &&
+				!idxStats.IsInvalid(coll.Pseudo) {
+				return true, idxStatsIdx
+			}
+		}
+	}
+	return false, -1
+}
+
+// GetSelectivityByFilter try to estimate selectivity of expressions by evaluate the expressions using TopN and NULL.
+// The data represented by the Histogram would use the defaultSelectivity parameter as the selectivity.
+// Currently, this method can only handle expressions involving a single column.
+func (coll *HistColl) GetSelectivityByFilter(sctx sessionctx.Context,
+	defaultSelectivity float64,
+	filters []expression.Expression) (ok bool, selectivity float64, err error) {
+	// 1. Make sure the expressions
+	//   (1) are safe to be evaluated here,
+	//   (2) involve only one column,
+	//   (3) and this column is not a "new collation" string column so that we're able to restore values from the stats.
+	for _, filter := range filters {
+		if expression.IsMutableEffectsExpr(filter) {
+			return false, 0, nil
+		}
+	}
+	if expression.ContainCorrelatedColumn(filters) {
+		return false, 0, nil
+	}
+	cols := expression.ExtractColumnsFromExpressions(nil, filters, nil)
+	if len(cols) != 1 {
+		return false, 0, nil
+	}
+	col := cols[0]
+	tp := col.RetType
+	if types.IsString(tp.GetType()) && collate.NewCollationEnabled() && !collate.IsBinCollation(tp.GetCollate()) {
+		return false, 0, nil
+	}
+
+	// 2. Get the available stats, make sure it's a ver2 stats and get the needed data structure from it.
+	isIndex, i := coll.findAvailableStatsForCol(sctx, col.UniqueID)
+	if i < 0 {
+		return false, 0, nil
+	}
+	var statsVer, nullCnt int64
+	var histTotalCnt, totalCnt float64
+	var topnTotalCnt uint64
+	var hist *Histogram
+	var topn *TopN
+	if isIndex {
+		stats := coll.Indices[i]
+		statsVer = stats.StatsVer
+		hist = &stats.Histogram
+		nullCnt = hist.NullCount
+		topn = stats.TopN
+	} else {
+		stats := coll.Columns[i]
+		statsVer = stats.StatsVer
+		hist = &stats.Histogram
+		nullCnt = hist.NullCount
+		topn = stats.TopN
+	}
+	// Only in stats ver2, we can assume that: TopN + Histogram + NULL == All data
+	if statsVer != Version2 {
+		return false, 0, nil
+	}
+	topnTotalCnt = topn.TotalCount()
+	histTotalCnt = hist.notNullCount()
+	totalCnt = float64(topnTotalCnt) + histTotalCnt + float64(nullCnt)
+
+	var topNSel, histSel, nullSel float64
+
+	// Prepare for evaluation.
+	originalIndex := col.Index
+	col.Index = 0
+	defer func() {
+		// Restore the original Index to avoid unexpected situation.
+		col.Index = originalIndex
+	}()
```

Comment:
What's the purpose for this operation? It seems the following part doesn't use the `col`.

Reply:
Please add this part to the comments.

Reply:
Updated.